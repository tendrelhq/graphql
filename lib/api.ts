import type { Connection } from "@/schema/system/pagination";
import { assert, type RawPaginationArgs, assertNonNull, map } from "./util";

type PaginationMeta = Omit<Connection<void>, "edges">;

export function extractPageInfo(res: Response): PaginationMeta {
  const contentRange = assertNonNull(
    res.headers.get("Content-Range"),
    "cannot paginate without content-range",
  );

  const [start, end, count] = map(contentRange.split("/"), cr =>
    [...cr[0].split("-"), cr[1]].map(Number),
  );
  assert(Number.isFinite(start));
  assert(Number.isFinite(end));
  assert(Number.isFinite(count));

  if (res.status === 206) {
    return {
      pageInfo: {
        hasNextPage: true,
        hasPreviousPage: false,
        endCursor: end.toString(),
        startCursor: start.toString(),
      },
      totalCount: count,
    };
  }

  return {
    pageInfo: {
      hasNextPage: false,
      hasPreviousPage: false,
    },
    totalCount: count,
  };
}

export function constructHeadersFromArgs(
  args: RawPaginationArgs,
  preferences?: string[],
): Headers {
  const headers = new Headers({
    Prefer: ["count=estimated"].concat(preferences ?? []).join(","),
  });
  const lower = Number(args.after ?? 0);
  const count = Math.min(args.first ?? 100, 100);
  const upper = lower + count - 1;
  headers.append("Range", `${lower}-${upper}`);
  return headers;
}
