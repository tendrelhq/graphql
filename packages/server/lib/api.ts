import type { Connection } from "@/schema/system/pagination";
import { assert, type RawPaginationArgs, assertNonNull, map } from "./util";

type PaginationMeta = Omit<Connection<void>, "edges">;

export function extractPageInfo(res: Response): PaginationMeta {
  const contentRange = assertNonNull(
    res.headers.get("Content-Range"),
    "cannot paginate without content-range",
  );

  const range = map(contentRange.split("/"), cr => {
    const [start, end] = cr[0].split("-");
    return {
      start: start,
      end: end,
      count: Number(cr[1]),
    };
  });
  assert(Number.isFinite(range.count), `invalid count: ${contentRange}`);

  if (range.start === "*" || res.status !== 206) {
    // Range not satisfiable
    return {
      pageInfo: {
        hasNextPage: false,
        hasPreviousPage: false,
      },
      totalCount: range.count,
    };
  }

  const start = Number(range.start);
  const end = Number(range.end);
  assert(Number.isFinite(start), `invalid start: ${contentRange}`);
  assert(Number.isFinite(end), `invalid end: ${contentRange}`);
  return {
    pageInfo: {
      hasNextPage: end < range.count - 1,
      hasPreviousPage: start > 0,
      endCursor: end.toString(),
      startCursor: start.toString(),
    },
    totalCount: range.count,
  };
}

export function constructHeadersFromArgs(
  args: RawPaginationArgs,
  preferences?: Record<string, string> & {
    count?: "estimated" | "exact" | "planned";
  },
): Headers {
  const prefers = preferences
    ? Object.keys(preferences)
        .map(key => `${key}=${preferences[key]}`)
        .join(",")
    : "count=estimated";
  const headers = new Headers({ Prefer: prefers });
  const lower = Number(args.after ?? 0);
  const count = Math.min(args.first ?? 100, 100);
  const upper = lower + count - 1;
  headers.append("Range-Unit", "items");
  headers.append("Range", `${lower}-${upper}`);
  return headers;
}
