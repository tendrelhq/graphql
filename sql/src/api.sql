CREATE SCHEMA api;
COMMENT ON SCHEMA api IS '
# Tendrel REST API

## Authentication and authorization

Most of the public api requires two tokens to be present in every request:

1. An application token, in JWT format via the X-Tendrel-App header.
2. An authorization token, in JWT format via the Authorization header.

The application token is essentially an OAuth 2.0 client secret, where the
"application" (e.g. a mobile app) is the OAuth 2.0 client. Application tokens
are required for all api calls.

The authorization token is a Tendrel issued JWT that uniquely identifies the
user making the request. Most api calls require a valid authorization token.
A client may exchange an IDP-issued security token for a Tendrel-issued token by
calling the /token api, which is implemented as an [OAuth 2.0 Token
Exchange](https://datatracker.ietf.org/doc/html/rfc8693).

### The "anonymous" role

The anonymous (`anon`) role can only be used to perform a select few operations
that do not require an authorization token. In particular, the `anon` role can
be used to signup for a Tendrel account as well as in various "introspection"
related operations, e.g. introspecting the OpenAPI schema. This role is the
default role for HTTP requests which do not include an authorization token.

## Localization

Tendrel is not a content management system. It does, however, provide mechanisms
that allow for content to be dynamically localized according to user preference.

Dynamic localization can be customized using various HTTP headers:

- `Accept-Language: en` specifies the locale.
- `Prefer: timezone=America/Denver` specifies the timezone.
';

