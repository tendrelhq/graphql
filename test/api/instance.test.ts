import { describe, expect, test } from "bun:test";
import { baseurl } from "@/test/api/constants";

describe("/api/instance", () => {
  test("create", async () => {
    const url = new URL("/instance?select=id", baseurl);
    const response = await fetch(url, {
      method: "POST",
      headers: {
        Prefer: "return=representation, tx=rollback",
      },
      body: JSON.stringify({
        name: "this is a test",
        owner: "70f200bd-1c92-481d-9f5c-e6cf6cd92cd0",
      }),
    });
    expect(response.json()).resolves.toEqual([
      {
        id: expect.any(String),
      },
    ]);
  });

  test("read", async () => {
    const url = new URL(
      "/instance?select=id,display_name(value),fields(id,display_name(value)),template(id,display_name(value))&order=id",
      baseurl,
    );
    const response = await fetch(url, {
      method: "GET",
      headers: {
        Range: "0-19",
      },
    });
    expect(response.json()).resolves.toMatchSnapshot();
  });

  test.skip("update", async () => {
    const template = "not-a-real-uuid"; // FIXME: need a demo customer
    const url = new URL(
      `/instance?id=eq.${template}&select=id,_order`,
      baseurl,
    );
    const response = await fetch(url, {
      method: "PATCH",
      headers: {
        Prefer: "return=representation",
      },
      body: JSON.stringify({ _order: 2 }),
    });
    expect(response.json()).resolves.toMatchSnapshot();
  });

  test("delete", async () => {
    const template = "0001570b-d98f-4617-9e21-d248c24dece5"; // FIXME: need a demo customer
    const url = new URL(
      `/instance?id=eq.${template}&select=id,_deleted`,
      baseurl,
    );
    const response = await fetch(url, {
      method: "DELETE",
      headers: {
        // Use tx=rollback here while we get the test suites up and running.
        // Eventually we should use a generated customer like we do in the
        // existing runtime test suites.
        Prefer: "return=representation, tx=rollback",
      },
    });
    expect(response.json()).resolves.toMatchSnapshot();
  });
});
