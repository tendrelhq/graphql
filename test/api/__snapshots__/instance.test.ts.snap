// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/instance read 1`] = `
[
  {
    "display_name": {
      "value": "System Tag",
    },
    "fields": [
      {
        "display_name": {
          "value": "systagabbreviationentityuuid",
        },
        "id": "7a9d3d7e-fd21-4352-8ecd-b7e2309f1a56",
      },
      {
        "display_name": {
          "value": "systagname",
        },
        "id": "139d072a-de1b-44c7-9038-e9e8d8e74c84",
      },
      {
        "display_name": {
          "value": "systagdisplayname",
        },
        "id": "8e1a82f3-0554-43a8-89f2-a70711c1aa72",
      },
      {
        "display_name": {
          "value": "systagtype",
        },
        "id": "9ca5a08e-63da-47d3-ab8c-f012d9c36758",
      },
    ],
    "id": "0001570b-d98f-4617-9e21-d248c24dece5",
    "template": {
      "display_name": {
        "value": "System Tag",
      },
      "id": "def05966-06b2-483e-8988-d0f898e45e6c",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "50f96456-573f-4d9e-bc04-21a63c09a38b",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "6d1e3dea-bb3a-427e-97ed-65b240d0a0c5",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "860371a8-74ce-4fa0-8fef-fe8c88422f2c",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "0ecd88b8-44c6-42e2-b91b-68796159a5cf",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "6b96e4c0-9898-4a9b-bee0-c3bd7e1a080e",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "26fb443e-0501-4ab3-bf4e-f9e8e77f6519",
      },
    ],
    "id": "000307a2-8b1e-4e4c-bc6a-e59dcf043cc4",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "cd38b2ff-9bc5-4a0f-8103-dae707fea409",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "1e74f21c-6093-43b6-afb4-b36808977cc5",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "fff4f973-dc35-4fdd-a956-cb09ff36c4d0",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "408c69a5-8515-4e2b-bb11-2cadaa58ff53",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "f0e5c968-0bb4-4414-85c2-8a28bb0afcb1",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "d4063acb-0e96-48b5-aff8-5f469460db4c",
      },
    ],
    "id": "00037232-f2cf-46d5-bf5f-3626dedfcf36",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "88488bbe-7e75-4f89-819c-cbf5f77d8b4f",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "3a8d4999-a654-464c-9060-be1c13812eb7",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "58874e50-5552-4c57-888f-bf8aabfeb21d",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "3d6de6a6-b937-4dcc-95d9-ff944f82e4ab",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "4c41e01b-46d0-4675-8b17-86a62b53a231",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "ffc8c5e8-4595-466d-a597-60d4d3ba343c",
      },
    ],
    "id": "00097e1d-7773-414f-b9c5-0fc5784e804a",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "ca1ea18e-0b4c-467d-b345-6a8abc18dd49",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "5426f2df-7e17-4111-805b-975496e51d9b",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "aafc9e48-8b3f-4ec1-ae46-831ed0dd8091",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "1a81744f-f2ef-4787-ab6a-7fd246b960e4",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "16b91352-815a-4c55-aa96-e528d8de728f",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "fa2c6267-8780-451b-959c-2bea2df89ab9",
      },
    ],
    "id": "00108a77-3163-4f88-9987-4d72883ad737",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "565e5c53-02a6-49f4-a0ee-2dad1f766f93",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "0f228a2f-33f1-4174-a3b8-8b6cfc0ec96d",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "281e3073-c2cb-4966-8147-1df2adb0734e",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "257246e3-3efb-4be3-96db-db1ec8847357",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "036a4046-cd12-4e8e-ac6e-0812dac2a3fd",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "a7859bec-45a4-4c78-939c-e308165cb276",
      },
    ],
    "id": "0012af66-9c1d-4cfb-9280-54d352232fd4",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "86daaa3c-1c34-4c69-b946-c38e7998db65",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "d8adee99-4d26-41df-8d1e-2e4d63106c4e",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "e1bf8f8b-c55e-420a-a6b8-09e7c6b18fab",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "68e20ac4-89e5-4df7-9dda-1a67cd3ec661",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "7c41736a-36f3-43db-befc-4f3af347e695",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "35fb5528-cf4a-49dc-9002-f3b74a65594d",
      },
    ],
    "id": "0014f5a1-284a-48ff-9047-605c1e569a8a",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "73bf8d08-004d-40c9-aab6-d773679a38d8",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "90ad9a6e-5e63-41a6-91aa-f8dab4b208ea",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "8857fa83-847a-45af-ad13-7d8f20557520",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "f625979f-9d7a-4c22-9e86-4bdbd81a106f",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "6a62d3ba-67ad-40ba-a86e-1e57157e27d2",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "a4d077cf-0d89-4086-9e82-e57d3a5b6049",
      },
    ],
    "id": "0015a019-8cab-4d37-a06a-1b0633023562",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "97c6af25-3cb4-49cb-8e2d-9b7d85d3b260",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "4c9c438b-36c5-46d4-a702-0a87bf8e43a2",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "8457e83a-91f7-4a12-a2cd-771b49954e79",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "fd948723-e9d1-4d65-af85-1f605b162a9c",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "91096a24-1ee8-4fd8-849f-bf7e892beef2",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "d0deb7e4-c29a-4c62-a964-65adfe6f6619",
      },
    ],
    "id": "001685e1-adfe-4d6a-86f0-4329e6f9261c",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "2d15d48d-4019-4f24-9f44-9cf1f6b6900c",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "611ecde0-9cc5-40dc-827e-dba218be3bc1",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "287a8821-d884-4da6-b154-2f138f153069",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "71cadfcd-d6b0-4fc7-abfb-d056a77e6bdb",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "38b03640-0731-4dec-8e2b-85fc7f8c4999",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "57c30d2b-dfca-4f68-8ee5-fe3825ff78b9",
      },
    ],
    "id": "0017d1bd-2c7e-46ba-b2f0-af3384d290ff",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "f39e45eb-c22a-4fe0-9bfe-36d0dc9f360d",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "a48e2615-6ef6-4ac1-8488-844823f07fe9",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "3508010a-5321-4fc6-861f-66fba9715a13",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "83189489-bccf-4924-bd5e-fa6820a20ede",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "6074e8d1-b90a-4fa2-ba17-08a32623c736",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "fa3731a7-d600-439d-bdbb-928a94118d14",
      },
    ],
    "id": "0018d21b-e9b4-418a-b96c-dac218665e15",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "acd12f09-0816-43fd-84a2-0ea9f3c62a23",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "980e9300-514e-4d97-a87d-ea47aea28d80",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "f96c3f38-3935-4364-8dd5-dd347a390506",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "31b634ab-76a3-4c41-ac9a-efb9f11598d1",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "6976a360-5f7a-4a59-a0ac-d64f34699178",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "ebfb59ae-c64e-4d98-8ff2-cbfc629323de",
      },
    ],
    "id": "001e9e0a-920a-460b-a7a2-50cb3476b745",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Customer Tag",
    },
    "fields": [
      {
        "display_name": {
          "value": "custagname",
        },
        "id": "54eeaa56-1db2-4c31-95c4-96cf8a2d2ad5",
      },
      {
        "display_name": {
          "value": "custagdisplayname",
        },
        "id": "197faa71-20b9-47eb-8b6f-10c08a763166",
      },
      {
        "display_name": {
          "value": " custagtype",
        },
        "id": "7d595577-562c-4d14-9fa1-01f40138dcc5",
      },
      {
        "display_name": {
          "value": "custagabbreviationentityuuid",
        },
        "id": "fe6d5b2e-698e-4000-8fae-faed63a0c835",
      },
    ],
    "id": "0022dff7-5211-4ae6-8bfd-c405d1d3b59b",
    "template": {
      "display_name": {
        "value": "Customer Tag",
      },
      "id": "30a317b8-6a56-45b4-8480-f9b58e099c77",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "e4226391-7279-4274-b427-cdf6931fe106",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "dea73d06-6cc2-473c-bfba-4252928f7bae",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "35a821ae-be64-4925-8974-d95a0c7996df",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "952139eb-e586-4eae-8e57-188faa49d302",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "1017c49c-2ab1-405e-80a2-cb929a86bf7b",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "02f6a3ff-b888-41e3-b8ba-4cde024cc62b",
      },
    ],
    "id": "00236bf1-e97f-4d0d-bcd9-5505243f6bb1",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "912b510c-7c48-4e41-bf5e-e9003ee55788",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "0a7c55b4-a9f9-45b5-ac9e-770e056ba56b",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "297cfd38-10f7-46d6-84d1-b5e9b495f2f1",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "7c095f05-1066-4dda-8595-1c217bcfc2f6",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "7da04010-9eb9-4deb-999a-aabd759fae2a",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "4521503e-5a72-45f4-b56b-3d99eaefbbef",
      },
    ],
    "id": "0025ba48-115e-408e-8a63-6a2ce74a1462",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "a3ec5689-d159-4fdb-b181-9bb4027d1791",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "5e328c1a-96f3-4f50-a148-40422daa3f44",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "a2bf80ce-edf2-42b4-8363-a22ed4c289ba",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "fecb81af-3d2c-4395-9cc3-2396190c726d",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "14c26ea2-d83d-4a0e-9188-cde07b404e3e",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "45047b46-f1b0-407f-96f9-e8a0012f9a5e",
      },
    ],
    "id": "0027c57d-6ca7-4047-a651-01798a678aba",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "c9a24fd7-1766-4949-b509-3432fbfba32f",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "8496334e-b7d7-4d8d-a08a-9df6082121e8",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "1d0c530c-9b70-4a74-a843-1048410e8468",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "7ed0e57f-8dc7-4f9e-b7c7-7c91d6ae3842",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "e7603de3-43bb-47f1-9f74-c609bdc63826",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "cea7581f-12dc-4199-bb8b-baf61b5b7887",
      },
    ],
    "id": "002c5f34-8459-4361-8fb4-d6496d1fcf22",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "adaaf22f-a090-4ba9-81b4-4f5f9e86a5bb",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "f734d36e-143c-413e-b88d-3223e94444c9",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "bb89be0a-6e37-4c16-930d-24d4868dad82",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "9c0dfc4c-1355-4f55-b3bc-020a08773433",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "c44943e9-4121-409c-a264-89709597b10f",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "b828cada-97a0-4450-9a90-165567d9d949",
      },
    ],
    "id": "002cb258-6b55-4140-b0a8-d2d993e73416",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "0fe8428f-fdec-430e-8c11-81273010cfcd",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "ad152554-80a2-4b1a-aae7-90ff6c38170d",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "1bfd5986-f881-4911-b4c0-091ff5a6db70",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "2e49ecc2-6ace-43de-b1aa-43e45a990459",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "9738c4c0-9e0b-45eb-bb3f-36ae870c83db",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "39015476-b57e-44df-9ebe-635ccc5fe417",
      },
    ],
    "id": "002e97ad-1301-4335-9e70-3256072e92d5",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "921964ab-2506-4820-a8a5-cd2c05025632",
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "64809f9a-f80e-488d-9a14-6eaf7eecc0b7",
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "8f07d86a-31ed-4945-95de-7e89bc256863",
      },
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "0bd7e6a0-2b19-446d-8fa6-dd635bcf21f9",
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "8fce8b9a-18ee-46a5-b834-5233c7f804d5",
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "e5d848b5-ebb3-44e7-a9f0-a3c8d774a0c0",
      },
    ],
    "id": "002f1ed1-ef1a-4ac6-86e9-577ad7703497",
    "template": {
      "display_name": {
        "value": "Location",
      },
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  },
]
`;

exports[`/api/instance delete 1`] = `
[
  {
    "_deleted": false,
    "id": "0001570b-d98f-4617-9e21-d248c24dece5",
  },
]
`;
